Redesign CookieStorage and Implement Leave Secure Cookie Alone

nmvk opened the pull request:
CookieStorage has been refactored to use HashMap with the base domain as the key. Values of hashmap are vector of cookies.
CookieStorage now has max_per_host which restricts maximum cookies that can be added per base domain.
Cookie eviction does not take place if max_per_host is not reached.
Cookie eviction logic implemented here does following steps
1) Evict all expired cookies
2) Remove oldest accessed non-secure cookie If any
3) When no non-secure cookie exists, remove oldest accessed secure cookie if new cookie being added is secure. Else ignore new cookie


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14445)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/cookie_storage.rs

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -8,46 +8,59 @@
             version: 1,
-            cookies: Vec::new()
+            cookies_map: HashMap::new(),
+            max_per_host: 50,
```
We should make this an argument to the constructor to allow the tests to use smaller values. The value used by the non-test code should be 150 in order to match other browsers.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -8,46 +8,59 @@
 
     // http://tools.ietf.org/html/rfc6265#section-5.3
     pub fn remove(&mut self, cookie: &Cookie, source: CookieSource) -> Result<Option<Cookie>, ()> {
+        // Step 0
```
This code doesn't correspond to a step 0 in the specification, so we shouldn't add this comment.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -8,46 +8,59 @@
     pub fn remove(&mut self, cookie: &Cookie, source: CookieSource) -> Result<Option<Cookie>, ()> {
+        // Step 0
+        let domain = reg_host(&cookie.cookie.domain.clone().unwrap_or("".to_owned())).unwrap_or("".to_owned());
+        if self.cookies_map.contains_key(&domain) {
```
It's more idiomatic to write:
```rust
let cookies = self.cookies_map.entry(&domain).or_insert(vec![]);
```
which means that the rest of the method will not need to be modified, since it works whether there is an existing entry or not.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -116,11 +194,40 @@ impl CookieStorage {
         }))
     }
 }
+    fn reg_host<'a>(url: &'a str) -> Option<String> {
```
This can return a `String`, since it always returns `Some(..)`.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -116,11 +194,40 @@ impl CookieStorage {
+        } else {
+            return false;
+        }
+    }
```
nit: deindent check_cookie_expired and reg_host one level.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -8,46 +8,59 @@
     // http://tools.ietf.org/html/rfc6265#section-5.3
     pub fn remove(&mut self, cookie: &Cookie, source: CookieSource) -> Result<Option<Cookie>, ()> {
+        // Step 0
+        let domain = reg_host(&cookie.cookie.domain.clone().unwrap_or("".to_owned())).unwrap_or("".to_owned());
```
`let domain = reg_host(cookie.cookie.domain.as_ref().unwrap_or(""));`

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -8,46 +8,59 @@
-            c.cookie.domain == cookie.cookie.domain &&
-            c.cookie.path == cookie.cookie.path &&
+        let position = cookies.iter().position(|c| {
+            c.cookie.domain == cookie.cookie.domain && c.cookie.path == cookie.cookie.path &&
```
Let's keep the path check on a separate line.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
         // Step 12
-        self.cookies.push(cookie);
+        let domain = reg_host(&cookie.cookie.domain.clone().unwrap_or("".to_owned())).unwrap_or("".to_owned());
+        if !self.cookies_map.contains_key(&domain) {
```
Same as before - we can use:
```rust
let domain = reg_host(&cookie.cookie.domain.as_ref().unwrap_or(""));
let cookies = self.cookies_map.entry(&domain).or_insert(vec![]);
```

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+        }
+        let mut cookies = self.cookies_map.get_mut(&domain).unwrap();
+
+        if cookies.len() == self.max_per_host {
```
Could we move all of the eviction implementation into a `evict_one_cookie` helper method?

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+        let mut cookies = self.cookies_map.get_mut(&domain).unwrap();
+
+        if cookies.len() == self.max_per_host {
+            // Step 12.1
```
This code doesn't correspond with a step 12.1 from the spec, so this comment doesn't belong here.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+
+                        is_first = false;
+                    }
+                }
```
We can write this more idiomatically:
```rust
let oldest_accessed: Option<(usize, Tm)> = None;
for (i, c) in cookies {
    if !c.cookie.secure && oldest_accessed.as_ref().map_or(true, |a| c.last_access < a.1) {
        oldest_accessed = Some((i, c.last_access));
    }
}
```

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+                }
+
+                // All secure cookies were found
+                if is_first {
```
We can use:
```rust
if let Some((index, _)) {
    cookies.remove(index);
} else {
    ...
}
```

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+                if is_first {
+                        if !cookie.cookie.secure {
+                            return;
+                        } else {
```
No need for an else after a return.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+                            let mut index = 0;
+                            let mut acc_time = time::now();
+                            // Get secure cookie with the oldest access time
+                            for i in 0..cookies.len() {
```
We should use the same pattern here as my previous review comment. We could even add a helper method that took a boolean argument to compare against the cookie's security to avoid the code duplication.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -65,7 +78,61 @@ impl CookieStorage {
+                                }
+                            }
+                            cookies.remove(index);
+                        }
```
nit: the code inside this block is indented more than necessary.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -87,14 +154,25 @@ impl CookieStorage {
 
         // Step 2
-        let mut url_cookies: Vec<&mut Cookie> = self.cookies.iter_mut().filter(filterer).collect();
+        let domain = reg_host(&url.host_str().unwrap_or("").to_owned()).unwrap();
```
`let domain = reg_host(url.host_str().unwrap_or(""));`

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -87,14 +154,25 @@ impl CookieStorage {
         // Step 2
-        let mut url_cookies: Vec<&mut Cookie> = self.cookies.iter_mut().filter(filterer).collect();
+        let domain = reg_host(&url.host_str().unwrap_or("").to_owned()).unwrap();
+        let host = domain.clone();
```
This shouldn't be necessary.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -87,14 +154,25 @@ impl CookieStorage {
+            let cookies: Vec<Cookie> = Vec::new();
+            self.cookies_map.insert(domain, cookies);
+        }
+        let mut cookies = self.cookies_map.get_mut(&host).unwrap();
```
`let cookies = self.cookies_map.entry(&domain).or_insert(vec![]);`

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -116,11 +194,40 @@ impl CookieStorage {
+            let cookies: Vec<Cookie> = Vec::new();
+            self.cookies_map.insert(domain, cookies);
+        }
+        let mut cookies = self.cookies_map.get_mut(&host).unwrap();
```
Same as previous comments for reg_host and using the entry API.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -116,11 +194,40 @@ impl CookieStorage {
+       Some(reg_suffix(url).to_string())
+    }
+
+    fn check_cookie_expired(cookie: &Cookie) -> bool {
```
Let's call this `is_cookie_expired`.

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -116,11 +194,40 @@ impl CookieStorage {
+    }
+
+    fn check_cookie_expired(cookie: &Cookie) -> bool {
+        if let Some(cookie_expiry_time_) = cookie.expiry_time {
```
This function is equivalent to the following:
```rust
fn is_cookie_expired(cookie: &Cookie) -> bool {
    match cookie.expiry_time {
        Some(t) => t.to_timespec() <= time::get_time(),
        None => false,
    }
}
```

jdm commented on tests/unit/net/cookie.rs:
```diff
@@ -132,3 +133,160 @@ fn test_sort_order() {
 }
+
+
+fn run(set_location: &str, set_cookies: &[String], final_location: &str) -> String {
```
Let's give this a more meaningful name.

jdm commented on tests/unit/net/cookie.rs:
```diff
@@ -132,3 +133,160 @@ fn test_sort_order() {
+    // Add all cookies to the store
+    for str_cookie in set_cookies {
+        let bytes = str_cookie.to_string().into_bytes();
+        let header = Header::parse_header(&[bytes]);
```
We should use `unwrap()` here to verify that the headers are being parsed correctly.

jdm replied:
Instead of having `if let Ok(..)` on the next line (which would skip the code if the header did not parse correctly), we should use `let header = Header::parse_header(&[bytes]).unwrap();`, so the test will fail if the header does not parse correctly.

jdm commented on tests/unit/net/cookie.rs:
```diff
@@ -132,3 +133,160 @@ fn test_sort_order() {
+        let header = Header::parse_header(&[bytes]);
+        if let Ok(SetCookie(cookies)) = header {
+            for bare_cookie in cookies {
+                if let Some(cookie) = Cookie::new_wrapped(bare_cookie, &url, source) {
```
We should use `unwrap()` here to verify that the cookies are wrapped correctly.

jdm commented on tests/unit/net/cookie.rs:
```diff
@@ -132,3 +133,160 @@ fn test_sort_order() {
+    for i in 1..47 {
+        let mut st: String = "extra".to_owned();
+        st.push_str(&i.to_string());
+        st.push_str(&"=bar; Secure; expires=Sun, 18-Apr-2000 21:06:29 GMT");
```
This could be `let st = format!("extra{}=bar; Secure; expires=Sun, 18-Apr-2000 21:06:29 GMT", i);`

jdm commented on tests/unit/net/cookie.rs:
```diff
@@ -132,3 +133,160 @@ fn test_sort_order() {
+    for i in 1..47 {
+        let mut st: String = "extra".to_owned();
+        st.push_str(&i.to_string());
+        st.push_str(&"=bar; Secure; expires=Sun, 18-Apr-2027 21:06:29 GMT");
```
Same comment about `format!()`.

jdm commented on tests/unit/net/cookie.rs:
```diff
@@ -132,3 +133,160 @@ fn test_sort_order() {
+                extra31=bar; extra32=bar; extra33=bar; extra34=bar; extra35=bar; extra36=bar; \
+                extra37=bar; extra38=bar; extra39=bar; extra40=bar; extra41=bar; extra42=bar; \
+                extra43=bar; extra44=bar; extra45=bar; extra46=bar; extra47=bar; extra48=bar; \
+                extra49=bar; extra50=bar");
```
These tests would be easier to read if we constructed a cookiestorage that only stored 5 cookies.

jdm reviewed (changes requested):
Great work! I've suggested some changes that will make the code easier to read, but this looks really good!

mrnayak commented:
@jdm Code review comments have been incorporated, Please let me know if there are any changes

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -78,57 +75,18 @@ impl CookieStorage {
-            self.cookies_map.insert(tempdomain, cookies);
-        }
-        let mut cookies = self.cookies_map.get_mut(&domain).unwrap();
+        let domain = reg_host(&cookie.cookie.domain.as_ref().unwrap_or(&"".to_string()));
```
Does `unwrap_or("")` work here?

mrnayak replied:
We got the error when we used `unwrap_or("")` expected String reference found str. So we used `&"".to_string()`

jdm commented on components/net/cookie_storage.rs:
```diff
@@ -78,57 +75,18 @@ impl CookieStorage {
-                        }
-                } else {
-                    cookies.remove(index);
+                if !evict_one_cookie(cookie.cookie.secure, cookies) {
```
```rust
if new_len == old_len && !evict_one_cookie(cookie.cookie.secure, cookies) {
    return;
}
```

jdm reviewed (changes requested):
Wonderful! One question and one suggestion; otherwise, this looks ready to merge!

mrnayak commented:
Thanks @jdm , I have committed changes.

jdm commented:
If you squash the changes into a single commit I'll merge them!

nmvk committed:
Redesign CookieStorage and Implement Leave Secure Cookie Alone

CookieStorage has been refactored to use HashMap with base domain as the
key. Values of hashmap is vector of cookies.
CookieStorage now has max_per_host which restricts maximum cookies that
can be added per base domain.
Cookie eviction doesnot take place if max_per_host is not reached.
Cookie eviction logic implemented here does following steps
1) Evict all expired cookies
2) Remove oldest accessed non-secure cookie If any
3) When no non-secure cookie exist, remove oldest accessed secure cookie
if new cookie being added is secure. Else ignore new cookie

mrnayak commented:
@jdm Squashed two commits into one. Thank you!

KiChjang commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit c1518ad has been approved by `jdm`

<!-- @bors-servo r=jdm c1518adba86a3124c950fdc2399865d73afea8c4 -->

bors-servo commented:
:hourglass: Testing commit c1518ad with merge b05c27c...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/4509), [arm64](http://build.servo.org/builders/arm64/builds/4517), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4599), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1276), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1296), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4575), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4597), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1287), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1300), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2541)
